### PR TITLE
updated

### PR DESCRIPTION
**Pull request by rudrapratap63**

### CODE REVIEW

File: Constant.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+public class Constant {
```
Comment (chandni-codeup) [inline Constant.java:3]: file commentry is missing

File: CustomStack.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+// Create custom stack class for stack data structure
```
Comment (chandni-codeup) [inline CustomStack.java]: improper commentry

File: CustomStack.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    // Constructor to initialize the stack with a specific size
+    public CustomStack(int size) {
+        stack = new Game2048.GameState[size];  
+        top = -1;  
+        maxSize = size;
+    }
+
+    // Push method to add an item to the stack
+    public void push(Game2048.GameState state) {
```
Comment (chandni-codeup) [inline CustomStack.java]: push ?

File: CustomStack.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+// Create custom stack class for stack data structure
+// for
```
Comment (chandni-codeup) [inline CustomStack.java]: No documentation

File: CustomStack.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
+// Create custom stack class for stack data structure
+// for 
+
+public class CustomStack {
+    private final Game2048.GameState[] stack;
```
Comment (chandni-codeup) [inline CustomStack.java]: Why? -> Game248 number cant be there in the class name

File: Constant.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    public String INVALID_GRID_SIZE = "Invalid size. Defaulting to 4x4.";
+    public String INVALID_INPUT = "Invalid Input";
+    public String NEW_GAME = "New Game";
+    public String FONT_NAME = "SansSerif";
+    public String GAME_OVER_WITH_SCORE = "Game Over! Your score: ";
+    public String WANT_TO_PLAY_AGAIN = "\nDo you want to play again?";
+    public String GAME_OVER = "Game Over";
+    public String SOUND_FILE_PATH =  "./button_sound.wav";
+    public String STACK_OVERFLOW_ERROR = "Stack Overflow: Cannot push more elements.";
+    public static String SOUND_NOT_FOUND = "Sound file not found!";
```
Comment (chandni-codeup) [inline Constant.java]: why static ?

File: CustomStack.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
+// Create custom stack class for stack data structure
+// for 
+
+public class CustomStack {
+    private final Game2048.GameState[] stack;
+    private int top;
+    private final int maxSize;
+    Constant constant = new Constant();
+
+    // Constructor to initialize the stack with a specific size
```
Comment (chandni-codeup) [inline CustomStack.java:7]: why use constructor? not allowed .

File: CustomStack.java (hunk 3 above)
Comment (chandni-codeup) [inline CustomStack.java]: always make an instance for class then use, don't directly use class

File: CustomStack.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    public CustomStack(int size) {
+        stack = new Game2048.GameState[size];  
+        top = -1;  
+        maxSize = size;
+    }
+
+    // Push method to add an item to the stack
+    public void push(Game2048.GameState state) {
+        if (top < maxSize - 1) {
+            stack[++top] = state;
```
Comment (chandni-codeup) [inline CustomStack.java]: .not a good approach to directly increment/decrement -> fir increment before  then use

File: CustomStack.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        } 
+    }
+
+    // Pop method to remove and return the top item from the stack
+    public Game2048.GameState pop() {
+        if (!isEmpty()) {
+            return stack[top--];
```
Comment (chandni-codeup) [inline CustomStack.java]: .not a good approach to directly increment/decrement -> fir decrement before then use

File: CustomStack.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        } else {
+            return null; 
+        }
+    }
+
```
Comment (chandni-codeup) [inline CustomStack.java:31]: avoid unnecessary spacing

File: Game2048.java [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+import java.awt.event.KeyEvent;
+import javax.swing.BorderFactory;
+import javax.swing.JButton;
+import javax.swing.JFrame;
+import javax.swing.JOptionPane;
+import javax.swing.JPanel;
+
+public class Game2048 {
+    class GameState {
+        int[][] board;
```
Comment (chandni-codeup) [inline Game2048.java]: array name must be plural

File: Game2048.java [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+        int score;
+        GameState(int[][] board, int score) {
+            this.board = board;
+            this.score = score;
+        }
+    }
+    private int size;
+    private int[][] board;
+    private boolean moved = false;
```
Comment (chandni-codeup) [inline Game2048.java]: naming is inappropriate -> 'moved'

File: Game2048.java [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+    public static void main(String[] args) {
+        Game2048 game = new Game2048();
+        game.start();
+    }
+
+    // Save Previous State in 2048 Game
+    private void saveState() {
+        int[][] boardCopy = new int[size][size];
+        for (int i = 0; i < size; i++) {
+            System.arraycopy(board[i], 0, boardCopy[i], 0, size);
```
Comment (chandni-codeup) [inline Game2048.java]: don't use inbuilt methods

File: Game2048.java [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+        }
+        history.push(new GameState(boardCopy, score));
```
Comment (chandni-codeup) [inline Game2048.java]: use instance of the class!!!!!!!

File: Game2048.java [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+            }
+        }
+    }
+    
+    // Start function for setup JFrame and GUI
+    public void start() {
+        JFrame frame = new JFrame(constant.APP_TITLE);
+        String inputSize = JOptionPane.showInputDialog(frame, constant.ENTER_GRID_SIZE, constant.GRID_SIZE_TITLE,
+                JOptionPane.QUESTION_MESSAGE);
+        try {
```
Comment (chandni-codeup) [inline Game2048.java]: mention why you are using try catch, describe exception

File: Game2048.java [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+                size = 4;
+            }
+        } catch (NumberFormatException exception) {
+            JOptionPane.showMessageDialog(frame, constant.INVALID_GRID_SIZE, constant.INVALID_INPUT,
+                    JOptionPane.WARNING_MESSAGE);
+            size = 4;
+        }
+        board = new int[size][size];
+        addTile();
+        addTile();
```
Comment (chandni-codeup) [inline Game2048.java]: ??

File: Game2048.java [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+        newGameButton.setBorderPainted(false);
+        newGameButton.setBorder(BorderFactory.createCompoundBorder(
+        BorderFactory.createLineBorder(new Color(0x8f7a66), 2, true), 
+        BorderFactory.createEmptyBorder(10, 20, 10, 20)));
+
+        newGameButton.addMouseListener(new java.awt.event.MouseAdapter() {
+            @Override
+            public void mouseEntered(java.awt.event.MouseEvent evt) {
+                newGameButton.setBackground(new Color(0xa68b79)); 
+            }
```
Comment (chandni-codeup) [inline Game2048.java]: not an optimized approach

File: Game2048.java [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+            @Override
+            public void mouseExited(java.awt.event.MouseEvent evt) {
+                newGameButton.setBackground(new Color(0x8f7a66));
+            }
+        });
+        newGameButton.setFocusable(false);
+        JPanel buttonPanel = new JPanel();
+        buttonPanel.add(newGameButton);
```
Comment (chandni-codeup) [inline Game2048.java]: variable name!!!!!

File: Game2048.java [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+        frame.add(buttonPanel, BorderLayout.NORTH);
+        gamePanel = new GamePanel(board, size, score, highScore);
+        frame.add(gamePanel);
+        frame.setSize(500, 650);
+        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+        frame.setVisible(true);
```
Comment (chandni-codeup) [inline Game2048.java]: You could have made a method for this

File: Game2048.java [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+        frame.setAlwaysOnTop(true);
+        frame.setResizable(false);
+        frame.setIconImage(Toolkit.getDefaultToolkit().getImage(constant.IMAGE_ICON));
+        frame.addKeyListener(new KeyAdapter() {
+            
+            // Invoke Up, Down, Left, Right function based on keys
+            // Input - KeyEvent
```
Comment (chandni-codeup) [inline Game2048.java]: INDENTATATION!!!!

File: Game2048.java [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+                    case KeyEvent.VK_W, KeyEvent.VK_UP -> moveUp();
+                    case KeyEvent.VK_A, KeyEvent.VK_LEFT -> moveLeft();
+                    case KeyEvent.VK_S, KeyEvent.VK_DOWN -> moveDown();
+                    case KeyEvent.VK_D, KeyEvent.VK_RIGHT -> moveRight();
+                    case KeyEvent.VK_Z -> undo();
+                }
+                if (moved) {
+                    addTile();
+                    if (isGameOver()) {
+                        highScore = Math.max(score, highScore);
```
Comment (chandni-codeup) [inline Game2048.java]: No use In-Build method or classes i.e. Math , max()

File: Game2048.java [hunk 22]
```diff
@@ -0,0 +1,355 @@
...
+                                constant.GAME_OVER_WITH_SCORE + score + constant.WANT_TO_PLAY_AGAIN, constant.GAME_OVER,
+                                JOptionPane.YES_NO_OPTION);
+                        if (response == JOptionPane.YES_OPTION) {
+                            resetGame();
+                        } else {
+                            System.exit(0);
+                        }
+                    }
+                    gamePanel.updateBoard(board, score, highScore);
+                }
```
Comment (chandni-codeup) [inline Game2048.java]: INDENTATION can be good

File: Game2048.java [hunk 23]
```diff
@@ -0,0 +1,355 @@
...
+
+    // isGameOver function for checking is game over or not
+    // return Type - boolean
+    private boolean isGameOver() {
+        for (int i = 0; i < size; i++) {
+            for (int j = 0; j < size; j++) {
+                if (board[i][j] == 0 ||
+                        (i > 0 && board[i][j] == board[i - 1][j]) ||
+                        (i < size - 1 && board[i][j] == board[i + 1][j]) ||
+                        (j > 0 && board[i][j] == board[i][j - 1]) ||
```
Comment (chandni-codeup) [inline Game2048.java]: ??

File: Game2048.java [hunk 24]
```diff
@@ -0,0 +1,355 @@
...
+                    } else {
+                        board[target][j] = lastValue; 
+                        lastValue = board[i][j]; 
+                        if (target != i) {
+                            moved = true;
+                        }
+                        target++;
+                    }
+                    board[i][j] = 0; 
+                }
```
Comment (chandni-codeup) [inline Game2048.java]: ??

File: Game2048.java [hunk 25]
```diff
@@ -0,0 +1,355 @@
...
+            for (int j = 0; j < size; j++) {
+                if (board[i][j] == 0) {
+                    emptyCount++;
+                }
+            }
+        }
+        if (emptyCount == 0) {
+            return;
+        }
+        int randomPosition = (int) (Math.random() * emptyCount);
```
Comment (chandni-codeup) [inline Game2048.java]: In-build class and methods should not be used

File: Game2048.java [hunk 26]
```diff
@@ -0,0 +1,355 @@
...
+        int emptyIndex = 0;
+        for (int i = 0; i < size; i++) {
+            for (int j = 0; j < size; j++) {
+                if (board[i][j] == 0) {
+                    if (emptyIndex == randomPosition) {
+                        board[i][j] = (Math.random() < 0.9) ? 2 : 4;
```
Comment (chandni-codeup) [inline Game2048.java]: Again MATH , Random

File: GamePanel.java [hunk 27]
```diff
@@ -0,0 +1,119 @@
...
+    protected void paintComponent(Graphics graphics) {
+        super.paintComponent(graphics);
+        Graphics2D graphics2D = (Graphics2D) graphics;
+        graphics2D.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
+        drawBoard(graphics2D);
+        drawScore(graphics2D);
+    }
+
+    // Drawing Board and setColor and font size 
+    private void drawBoard(Graphics2D graphics2D) {
```
Comment (chandni-codeup) [inline GamePanel.java:39]: Improper comment

File: GamePanel.java [hunk 28]
```diff
@@ -0,0 +1,119 @@
...
+                drawTile(graphics2D, i, j, tileSize, padding);
+            }
+        }
+    }
+
+    // Drawing Tile component 
+    // Input - Graphics2d class, Integer
+    private void drawTile(Graphics2D graphics2D, int row, int column, int tileSize, int padding) {
+        int value = board[row][column];
+        int x = column * (tileSize + padding) + padding;
```
Comment (chandni-codeup) [inline GamePanel.java]: Variable name -> x,y????

File: GamePanel.java [hunk 29]
```diff
@@ -0,0 +1,119 @@
...
+    private void drawScore(Graphics2D graphics2D) {
+        graphics2D.setColor(new Color(0xbbada0));
+        graphics2D.fillRect(0, getHeight() - 50, getWidth(), 50); 
+        graphics2D.setColor(Color.WHITE);
+        graphics2D.setFont(new Font(constant.FONT_NAME, Font.BOLD, 28));
+        graphics2D.drawString("Score: " + score, 10, getHeight() - 15);
+        graphics2D.drawString(constant.BEST + highScore, getWidth() - graphics2D.getFontMetrics().stringWidth(constant.BEST + highScore) - 10, getHeight() - 15);
+    }
+
+    private Color getTileColor(int value) {
```
Comment (chandni-codeup) [inline GamePanel.java:88]: Method  Comments

File: SoundPlayer.java [hunk 30]
```diff
@@ -0,0 +1,24 @@
+// Sound Player for 2048 Game
```
Comment (chandni-codeup) [inline SoundPlayer.java]: improper commentry

File: SoundPlayer.java [hunk 31]
```diff
@@ -0,0 +1,24 @@
+// Sound Player for 2048 Game
+
+import java.net.URL;
+import javax.sound.sampled.AudioInputStream;
+import javax.sound.sampled.AudioSystem;
+import javax.sound.sampled.Clip;
+
+public class SoundPlayer {
+    public static void playSound(String fileName) {
+        try {
```
Comment (chandni-codeup) [inline SoundPlayer.java:16]: explain exception

**Review by chandni-codeup — CHANGES_REQUESTED** [review]

File: Constant.java (hunk 6 above)
Comment (rudrapratap63) [reply Constant.java]: because we can't print non static variable which is deifine in another class Constant

File: Game2048.java (hunk 11 above)
Comment (rudrapratap63) [reply Game2048.java]: because board is singular entity and i initialize only one 2d array which is named as board

**Pull request closed without merging**